<a href="https://colab.research.google.com/github/Datanyra/pandas_2nd_assessment/blob/main/Suliha_Assessment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **TS_ASSESSMENT**

###**Imports**


In [1]:
import numpy as np
import pandas as pd

from google.colab import files
uploaded = files.upload()

Saving saas.csv to saas (1).csv


In [2]:
df = pd.read_csv('saas.csv', dtype=str)
df

,account_id,company,signup_ts,country,seats,mrr,plan,last_active,churned
0,A-1001,Acme Corp,05/01/2023 09:00,US,25,2500,enterprise,01/03/2024,FALSE
1,A-1002,Beta LLC,14/02/2023,US,5,450,pro,02/03/2024,FALSE
2,A-1001,Acme Corp,05/01/2023 09:00,US,25,2500,enterprise,01/03/2024,0
3,A-1003,Cirrus,1675209600,GB,0,0,pro,28/02/2024,FALSE
4,A-1004,Delta Inc,20/03/2023,United States,12,1188,pro,01/03/2024,No
5,A-1005,Echo,01/04/2023,US,8,-99,pro,03/03/2024,FALSE
6,A-1006,Foxtrot,11/05/2023,us,3,270,PRO,02/03/2024,FALSE
7,A-1007,Gamma,30/06/2023,DE,15,1800,enterprise,15/06/2023,TRUE
8,A-1008,Helio,22/07/2023,US,NaN,540,pro,04/03/2024,FALSE
9,A-1009,Iris,08/08/2023,US,10,900,pro,NaN,TRUE


##**Section A: The Mental Model**

##### *Q1: Predict the alignment*



```
#my_prediction

a = pd.Series({'jan': 100, 'feb': 200, 'mar': 300})
b = pd.Series({'feb': 10, 'mar': 20, 'apr': 40})
result = a + b

output:
'jan' : NaN
'feb' : 210
'mar' : 320
'apr' : NaN
```



In [ ]:
a = pd.Series({'jan': 100, 'feb': 200, 'mar': 300})
b = pd.Series({'feb': 10, 'mar': 20, 'apr': 40})
result = a + b
print(result)

apr      NaN
feb    210.0
jan      NaN
mar    320.0
dtype: float64




```
my_explanation:
Pandas series are labelled arrays.
It usually aligns index labels rather than their positions in the series when performing arithmetic.
That is why it adds the values with matching index labels that are alike and returns NaN for index labels that do not have values in the other series.
The result has 4 rows because Pandas includes all unique index labels from both series.
Jan and apr have no matching index labels in the other series.
The mechanism responsible is INDEX ALIGNMENT
```



##### *Q2: Series or DataFrame?*


| Expression | Returned type | Consequence |
|------------|---------------|-------------|
| `df["mrr"]` | Series | Cannot select multiple columns like you can in a DataFrame. |
| `df[["mrr"]]` | DataFrame | Can select multiple columns from the result because it is a DataFrame. |
| `df.loc[:, "mrr"]` | Series | The `.columns` attribute is not available because the result is a Series. |
| `df.loc[:, ["mrr", "seats"]]` | DataFrame | Can select and work with multiple columns, also `.columns` attribute works because the result is a DataFrame. |




```

##**Section A: The Mental Model**

##**Section B: Column Slicing**

##### *Q3: Four routes to the same columns*

In [ ]:
#The 4 expressions
df[["company","seats","mrr"]]
df.loc[:,["company","seats","mrr"]]
df.iloc[:,[1,4,5]]
df.filter(items=["company","seats","mrr"])

,company,seats,mrr
0,Acme Corp,25,2500
1,Beta LLC,5,450
2,Acme Corp,25,2500
3,Cirrus,0,0
4,Delta Inc,12,1188
5,Echo,8,-99
6,Foxtrot,3,270
7,Gamma,15,1800
8,Helio,NaN,540
9,Iris,10,900


**Fragility Analysis**


The 2 fragile routes:

1. by_name(`df[["company","seats","mrr"]]`): This depends on naming, so if a colleague renames any of these columns, the code will fail cause the index labels don't exist anymore

2. by `.iloc` positions(`df.iloc[:,[1,4,5]]`): This depends on column position, i.e if a colleague reorders(adds,removes or changes) columns, the positions may change and return the wrong columns

**Production Choice with Justification**

I would put `.loc` in a production code because it is clear, easy understand,selects all rows and does not depend on columns staying in order,changes in column order does not affect result.

##### *Q4: The inclusive/exclusive trap*

*My Prediction*

This `df.loc[:, 'company':'mrr']` would print 4 columns:
`company`
`signup_ts`
`country`
`seats`

This `df.iloc[:, 1:5]` would print 4 columns also:
`company`
`signup_ts`
`country`
`seats`

In [ ]:
df.loc[:, 'company':'mrr']


,company,signup_ts,country,seats,mrr
0,Acme Corp,05/01/2023 09:00,US,25,2500
1,Beta LLC,14/02/2023,US,5,450
2,Acme Corp,05/01/2023 09:00,US,25,2500
3,Cirrus,1675209600,GB,0,0
4,Delta Inc,20/03/2023,United States,12,1188
5,Echo,01/04/2023,US,8,-99
6,Foxtrot,11/05/2023,us,3,270
7,Gamma,30/06/2023,DE,15,1800
8,Helio,22/07/2023,US,NaN,540
9,Iris,08/08/2023,US,10,900


In [ ]:
df.iloc[:, 1:5]

,company,signup_ts,country,seats
0,Acme Corp,05/01/2023 09:00,US,25
1,Beta LLC,14/02/2023,US,5
2,Acme Corp,05/01/2023 09:00,US,25
3,Cirrus,1675209600,GB,0
4,Delta Inc,20/03/2023,United States,12
5,Echo,01/04/2023,US,8
6,Foxtrot,11/05/2023,us,3
7,Gamma,30/06/2023,DE,15
8,Helio,22/07/2023,US,NaN
9,Iris,08/08/2023,US,10


**Rule Explanation**

* For `.loc` slicing is inclusive, it includes both the starting and ending label. I got caught in the trap apt.

* For `iloc` slicing is exclusive at the end, includes starting but excludes the the ending label.


##**Section C: Diagnosis and Cleaning**

##### *Q5: The count is the question*

In [ ]:
pd.DataFrame(df)

,account_id,company,signup_ts,country,seats,mrr,plan,last_active,churned
0,A-1001,Acme Corp,05/01/2023 09:00,US,25,2500,enterprise,01/03/2024,FALSE
1,A-1002,Beta LLC,14/02/2023,US,5,450,pro,02/03/2024,FALSE
2,A-1001,Acme Corp,05/01/2023 09:00,US,25,2500,enterprise,01/03/2024,0
3,A-1003,Cirrus,1675209600,GB,0,0,pro,28/02/2024,FALSE
4,A-1004,Delta Inc,20/03/2023,United States,12,1188,pro,01/03/2024,No
5,A-1005,Echo,01/04/2023,US,8,-99,pro,03/03/2024,FALSE
6,A-1006,Foxtrot,11/05/2023,us,3,270,PRO,02/03/2024,FALSE
7,A-1007,Gamma,30/06/2023,DE,15,1800,enterprise,15/06/2023,TRUE
8,A-1008,Helio,22/07/2023,US,NaN,540,pro,04/03/2024,FALSE
9,A-1009,Iris,08/08/2023,US,10,900,pro,NaN,TRUE


| Problem | Classification | Fix / No Fix | Why |
|----------|----------------|--------------|-----|
| Account ID `A-1001` appears twice | What is one row? | No Fix | I would not remove it immediately because I do not know if it is a duplicate or an update to the same account. I would confirm before making changes. |
| Account ID `A-1002` appears twice with different values | What is one row? | No Fix | The rows are different, so removing one could delete useful information. I would first check if one is an updated record. |
| `1675209600` in `signup_ts` | What should each column be? | Fix | The column should contain dates in one consistent format. I would convert the Unix timestamp to a date. |
| `United States` and `US` in `country` | What should each column be? | Fix | Both values mean the same thing, so I would make them consistent. |
| `PRO` and `pro` in `plan` | What should each column be? | Fix | They represent the same plan but use different capitalization. I would standardize them. |
| `False`, `FALSE`, `false`, `0`, and `No` in `churned` | What should each column be? | Fix | The column should use one consistent Boolean format. |
| Blank company name | What is missing or lying? | Fix | The company name is missing, so I would investigate or fill it if the correct value is available. |
| Missing `last_active` value | What is missing or lying? | Fix | Missing dates can affect analysis, so I would investigate why it is missing. |
| `mrr = -99` | What is missing or lying? | Fix | `-99` looks like a placeholder instead of a real revenue value, so I would confirm and replace it if necessary. |
| `last_active = 2099-01-01` | What is missing or lying? | Fix | Although it is a valid date, it looks unrealistic for this dataset and should be checked. |
| `seats = 1000000` | What is missing or lying? | No Fix | It looks unusual, but it could be a valid enterprise customer. I would verify it before changing it. |
| `seats = 0` and `mrr = 0` | What is missing or lying? | No Fix | I would leave it because it could be a free or inactive account. The data alone does not prove it is wrong. |

##### *Q6: The duplicate that isn't*

my_prediction to `df = df.drop_duplicates()`

It'd delete all repeated accounts rows

In [ ]:
df = df.drop_duplicates()
df
#No rows were deleted/removed because none of the repeated account rows were identical accross all columns

,account_id,company,signup_ts,country,seats,mrr,plan,last_active,churned
0,A-1001,Acme Corp,05/01/2023 09:00,US,25,2500,enterprise,01/03/2024,FALSE
1,A-1002,Beta LLC,14/02/2023,US,5,450,pro,02/03/2024,FALSE
2,A-1001,Acme Corp,05/01/2023 09:00,US,25,2500,enterprise,01/03/2024,0
3,A-1003,Cirrus,1675209600,GB,0,0,pro,28/02/2024,FALSE
4,A-1004,Delta Inc,20/03/2023,United States,12,1188,pro,01/03/2024,No
5,A-1005,Echo,01/04/2023,US,8,-99,pro,03/03/2024,FALSE
6,A-1006,Foxtrot,11/05/2023,us,3,270,PRO,02/03/2024,FALSE
7,A-1007,Gamma,30/06/2023,DE,15,1800,enterprise,15/06/2023,TRUE
8,A-1008,Helio,22/07/2023,US,NaN,540,pro,04/03/2024,FALSE
9,A-1009,Iris,08/08/2023,US,10,900,pro,NaN,TRUE


I would say `drop_duplicate()` is the wrong tool because what it does delete only accounts or rows that are identical accross all columns.

In this dataset, the repeated account IDs have different values (different `seats`, `mrr`, or `last_active`), so pandas treats them as different rows even though they refer to the same account.

Replacement code

Instead of deleting rows immediately, I would first identify the repeated account IDs using `df[df.duplicated(subset=['account_id'], keep=False)]`. Afterwards, I'd either use `df = df.drop_duplicates(subset=['account_id'], keep='last')` or `df = df.drop_duplicates(subset=['account_id'], keep='first')`
depending on the business rule.


**Missing business information I need**

Before committing to the code, I would need to know which record should be kept when an account ID appears more than once. For example, should I keep the newest record, the oldest record, the one with the latest `last_active` date, or should the records be combined instead?.

##### *Q7: One column, several diseases*

In [ ]:
#computing revenue per seat
#I will be converting both columns to numbers before dividing because every column is stored as an object
df["mrr"] = pd.to_numeric(df["mrr"])
df["seats"] = pd.to_numeric(df["seats"])
df["revenue"] = df["mrr"]/ df["seats"]
df


,account_id,company,signup_ts,country,seats,mrr,plan,last_active,churned,revenue
0,A-1001,Acme Corp,05/01/2023 09:00,US,25.0,2500,enterprise,01/03/2024,FALSE,100.00000
1,A-1002,Beta LLC,14/02/2023,US,5.0,450,pro,02/03/2024,FALSE,90.00000
2,A-1001,Acme Corp,05/01/2023 09:00,US,25.0,2500,enterprise,01/03/2024,0,100.00000
3,A-1003,Cirrus,1675209600,GB,0.0,0,pro,28/02/2024,FALSE,NaN
4,A-1004,Delta Inc,20/03/2023,United States,12.0,1188,pro,01/03/2024,No,99.00000
5,A-1005,Echo,01/04/2023,US,8.0,-99,pro,03/03/2024,FALSE,-12.37500
6,A-1006,Foxtrot,11/05/2023,us,3.0,270,PRO,02/03/2024,FALSE,90.00000
7,A-1007,Gamma,30/06/2023,DE,15.0,1800,enterprise,15/06/2023,TRUE,120.00000
8,A-1008,Helio,22/07/2023,US,NaN,540,pro,04/03/2024,FALSE,NaN
9,A-1009,Iris,08/08/2023,US,10.0,900,pro,NaN,TRUE,90.00000


Most of the rows have a revenue per seat of **90**, so I used **90** as my baseline. Since the majority of the accounts cluster around this value, it appears to represent the normal or expected revenue per seat. Any row that differs significantly from 90 deserves further investigation, but not every difference automatically means the data is wrong. Some differences may represent valid business cases while others are genuine data quality issues.

| Account ID | Revenue per seat | Most likely explanation | Fix |
|------------|-----------------:|--------------------------|-----|
| A1001 | 100 | Slightly higher than the baseline. This could simply be a different pricing plan or a premium customer. There is not enough evidence that it is incorrect. | No immediate fix. Verify with pricing rules if necessary. |
| A1003 | NaN | Seats are 0, so revenue per seat cannot be calculated because division by zero is undefined. This may be a legitimate account with no active seats rather than dirty data. | Leave it until the business confirms whether 0 seats is valid. |
| A1005 | -12.375 | Negative revenue per seat is suspicious. The value **-99** is likely a placeholder code for a missing MRR rather than a real revenue value. | Replace the placeholder with `NaN` and obtain the correct MRR if available. |
| A1007 | 120 | Higher than the baseline. This could either be a premium pricing plan or an incorrect seat/MRR value. The data alone cannot determine which explanation is correct. | Do not change the data until the business confirms the expected pricing. |
| A1008 | NaN | The number of seats is missing, so revenue per seat cannot be calculated. | Recover the missing seat value or leave it as missing until it can be verified. |
| A1010 | 0 | Revenue per seat is 0 because MRR is 0. This could mean the customer is on a free plan or that the MRR was entered incorrectly. | Confirm with the business whether a zero MRR is expected before making changes. |
| A1012 | 0.00009 | The seat count of **1,000,000** is extremely unlikely compared with the rest of the dataset. This is most likely a data entry error. | Verify the seat count against the source system and correct it if necessary. |

**Row with two irreconcilable readings**
`A1007 (Revenue per seat = 120)`

There are two equally reasonable explanations:

The customer is paying more because they are on a premium or enterprise pricing plan.
Either the MRR or the number of seats was entered incorrectly.
The dataset alone cannot distinguish between these explanations. To decide which is correct, I would need the company's pricing policy or subscription records showing the expected revenue per seat for that customer's plan.

**Why I did not include `A1004`**

Although A1004 has a revenue per seat of 99, I did not classify it as suspicious. It is still close to the baseline value of 90, and there is no evidence in the dataset that it is incorrect. A small difference in revenue per seat could simply be due to a different pricing agreement, discounts, taxes, or billing policy. Since there is no clear indication of a data quality problem, I would leave this row unchanged.

##### *Q8: Parsing is not validating*

In [ ]:
#code that flags both kinds

#code for parsing date columns
df["signup_ts"] = pd.to_datetime(df["signup_ts"],errors="coerce")
df["last_active"] = pd.to_datetime(df["last_active"],errors="coerce")
df


,account_id,company,signup_ts,country,seats,mrr,plan,last_active,churned
0,A-1001,Acme Corp,2023-05-01 09:00:00,US,25,2500,enterprise,2024-01-03,FALSE
1,A-1002,Beta LLC,NaT,US,5,450,pro,2024-02-03,FALSE
2,A-1001,Acme Corp,2023-05-01 09:00:00,US,25,2500,enterprise,2024-01-03,0
3,A-1003,Cirrus,NaT,GB,0,0,pro,NaT,FALSE
4,A-1004,Delta Inc,NaT,United States,12,1188,pro,2024-01-03,No
5,A-1005,Echo,NaT,US,8,-99,pro,2024-03-03,FALSE
6,A-1006,Foxtrot,NaT,us,3,270,PRO,2024-02-03,FALSE
7,A-1007,Gamma,NaT,DE,15,1800,enterprise,NaT,TRUE
8,A-1008,Helio,NaT,US,NaN,540,pro,2024-04-03,FALSE
9,A-1009,Iris,NaT,US,10,900,pro,NaT,TRUE


In [ ]:
#dates that could not be converted(parsing problems)
parsing_errors = df[df["signup_ts"].isna() | df["last_active"].isna()]
parsing_errors


,account_id,company,signup_ts,country,seats,mrr,plan,last_active,churned
3,A-1003,Cirrus,1675209600,GB,0,0,pro,NaT,FALSE
7,A-1007,Gamma,30/06/2023,DE,15,1800,enterprise,NaT,TRUE
9,A-1009,Iris,08/08/2023,US,10,900,pro,NaT,TRUE


In [ ]:
#validation problems(valid but makes no sense)
#future_dates = df[(df["signup_ts"] > pd.Timestamp.now()) | (df["last_active"] > pd.Timestamp.now())]
#future_dates
before_signup = df[df["last_active"] < df["signup_ts"]]
before_signup

,account_id,company,signup_ts,country,seats,mrr,plan,last_active,churned


##**SUMMARY**
**Code used to identify both kinds of problems**

*Parse the date columns*

`df["signup_ts"] = pd.to_datetime(df["signup_ts"], errors="coerce")`

`df["last_active"] = pd.to_datetime(df["last_active"], errors="coerce")`


*Parsing problems (dates that could not be converted)*

parsing_errors = `df[df["signup_ts"].isna() | df["last_active"].isna()]`

*Validation problems (dates that are valid but do not make business sense)*

future_dates = `df[(df["signup_ts"] > pd.Timestamp.now()) | (df["last_active"] > pd.Timestamp.now())]`

before_signup = `df[df["last_active"] < df["signup_ts"]]`

**Parsing problem:** A parsing problem happens when Pandas cannot convert a value into a date. In this case, pd.to_datetime() returns NaT, showing that the value could not be interpreted as a valid date.

*For example:*
The signup_ts column contains different formats, including standard dates and a Unix timestamp (1675209600). If these formats are not handled correctly during parsing, some values may become NaT.

**Validation problem:** A validation problem occurs after the date is successfully converted, but it still does not make sense based on the business rules.

*For example:*
A1013 has last_active = 2099-01-01. Pandas accepts this as a valid date, but it is unrealistic because it is far in the future.

A1007 has signup_ts = 30/06/2023 but last_active = 15/06/2023. Both dates are valid, but a customer cannot be active before signing up.

These rows show why parsing alone is not enough. A date can have the correct format but still represent impossible business data.

**Difference between parsing and validation**

Parsing checks whether a value can be converted into a date.

Validation checks whether the date makes sense in the real world or according to the business rules.

A date can pass parsing but still fail validation, which is why both checks are necessary during data cleaning.

##**Section D: Judgement**

##### *Q9: The same null, two fates*

**Label column with missing value**

Row: `A1014`

Missing column: `company`

I would not immediately drop this row. The missing value is only the company name, while the account ID, revenue, seats and other business information are still present. If I am calculating revenue or churn statistics, this row is still useful. I would first try to recover the company name from another source(maybe using the account ID) before deciding to remove it.


**Measure column with missing value**

Row: `A1008`

Missing column: `seats`

This missing value directly affects calculations such as revenue per seat. Since seats is a numeric measure used in analysis, I would not guess its value. I would either obtain the correct value from the source system or exclude this row only from analyses that require seats until the value is available.

**Why the same action is not appropriate**

Dropping both rows would remove useful business data unnecessarily.

Keeping both rows without considering the missing values could also produce incorrect analysis.

The correct decision depends on how the data will be used. For example, a revenue dashboard can still use `A1014` because its revenue is known, but a revenue per seat analysis cannot use `A1008` until the seats value is available or known.

##### *Q10: Cleaning is a function of the question*

**Decision 1: I decided not to remove the duplicate account IDs immediately.**

**Why I made that decision:**
The duplicate account IDs had different values, so I couldn't assume one of them was wrong. I felt it was better to investigate them first.

**When I would do the opposite:**
If I knew from the business that an account ID should only ever appear once, then I would remove the duplicates because I would know they were accidental duplicates.


**Decision 2: I decided to flag the impossible dates instead of changing them.**

**Why I made that decision:**
Although the dates could be read by pandas, I couldn't tell what the correct dates were, so changing them would just be guessing.

**When I would do the opposite:**
If the company already had a rule for correcting those dates, or I was given the correct values, then I would update them instead of only flagging them.



**Conclusion**

There isn't one version of a perfectly clean dataset. What counts as "clean" depends on what you're trying to do with the data, so the same cleaning decision can be right in one situation and wrong in another.